### PR TITLE
Fixes for the issues brought up by the Debian builds

### PR DESCRIPTION
**Pull request by nalinigans**
Note:

1. Remove embedded Catch2 header file and use `brew install catch2(v3.1.0)` for MacOS and `install from source release v2.13.9` for Linux. Note that both versions 2 and 3 are now supported.
2. Moving to the latest Catch2 has prompted the move to C++17 for `gcc`, `clang` was fine with c++11 btw, apart from test code changes.
3. C++17 caused a few issues with the `gcs-sdk` whose build had to be moved to C++17 as well.
4. Cleaned up caching in GitHub Actions to speedup CI builds.

### CODE REVIEW

File: .github/scripts/install_catch2.sh [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# THE SOFTWARE.
+#
+
+#!/bin/bash
+
+# Error out on first failure
+set -e
+
+INSTALL_DIR=${INSTALL_DIR:/usr/local}
+CATCH2_VER=${CATCH2_VER:v3.1.0}
```
Comment (webberlh) [inline .github/scripts/install_catch2.sh:33]: Looks like there's a missing `-` after `:` in setting the default value.

File: .github/scripts/install_catch2.sh [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+#!/bin/bash
+
+# Error out on first failure
+set -e
+
+INSTALL_DIR=${INSTALL_DIR:/usr/local}
```
Comment (webberlh) [inline .github/scripts/install_catch2.sh:32]: Could this default value be set to `/usr/local/catch2`? Also missing `-` after `:`.

File: .github/scripts/install_catch2.sh (hunk 1 above)
Comment (nalinigans) [reply .github/scripts/install_catch2.sh]: Thanks for catching this!

File: .github/scripts/install_catch2.sh (hunk 2 above)
Comment (nalinigans) [reply .github/scripts/install_catch2.sh]: catch2 installation defaults to `/usr/local` - `/usr/local/include/catch2`, `/usr/local/lib/cmake/Catch2`, etc. There is a `make uninstall` if we need to clean all the `Catch2` artifacts. Basically, I want to allow Catch2 to follow its defaults when installing into `/usr/local`.  Come to think of it, the default might be `/usr` in the case of Linux...

**Review by webberlh — APPROVED** [review]
Looks good to me!

**Pull request merged**